api/src/main/java/org/pmiops/workbench/billing/FreeTierBillingService.java

```diff
@@ -227,132 +360,200 @@ private void maybeAlertOnCostThresholds(
+   * may occur after the workspace gets deleted and after the last cron had run.
    *
-   * @param user the user as represented in our database
-   * @return the total USD amount spent in workspaces created by this user, represented as a double
+   * @param allCostsInDbForUsers
+   * @return a {@link java.util.List} of the workspaces that require updates only
    */
-  @Nullable
-  public Double getCachedFreeTierUsage(DbUser user) {
-    return workspaceFreeTierUsageDao.totalCostByUser(user);
+  @NotNull
+  private List<WorkspaceCostView> findActiveOrDeletedWorkspacesThatRequireUpdates(
+      List<WorkspaceCostView> allCostsInDbForUsers) {
+
+    List<WorkspaceCostView> workspacesThatRequireUpdate =
+        allCostsInDbForUsers.stream()
+            .filter(
+                c ->
+                    (c.getActiveStatus() == 0)
+                        || (c.getActiveStatus() == 1
+                            && (c.getFreeTierLastUpdated() == null
+                                || c.getWorkspaceLastUpdated() == null
+                                || c.getFreeTierLastUpdated().before(c.getWorkspaceLastUpdated())
+                                || (c.getFreeTierLastUpdated().after(c.getWorkspaceLastUpdated())
+                                    && c.getFreeTierLastUpdated().getTime()
+                                            - c.getWorkspaceLastUpdated().getTime()
+                                        < getNumberOfDaysToConsiderForFreeTierUsageUpdate()))))
+            .collect(Collectors.toList());
+
+    logger.info(
+        String.format("Workspaces that require update %d", workspacesThatRequireUpdate.size()));
+
+    return workspacesThatRequireUpdate;
   }
 
   /**
-   * Does this user have remaining free tier credits? Compare the user-specific free tier limit (may
-   * be the system default) to the amount they have used.
+   * Requesting live costs from BQ in batches then updating it in a transactional method.
    *
-   * @param user the user as represented in our database
-   * @return whether the user has remaining credits
+   * @param allCostsInDbForUsers
+   * @param workspacesThatRequireUpdate
+   * @return a Map of all live costs.
    */
-  public boolean userHasRemainingFreeTierCredits(DbUser user) {
-    final double usage = Optional.ofNullable(getCachedFreeTierUsage(user)).orElse(0.0);
-    return !costAboveLimit(user, usage);
+  private Map<Long, Double> updateFreeTierUsageInBatches(
+      List<WorkspaceCostView> allCostsInDbForUsers,
+      List<WorkspaceCostView> workspacesThatRequireUpdate) {
+    final Map<Long, Double> liveCostsInBQ = new HashMap<>();
+    for (List<WorkspaceCostView> workspacesPartition :
+        Lists.partition(workspacesThatRequireUpdate, 1000)) {
+      // Live cost in BQ
+      liveCostsInBQ.putAll(getFreeTierWorkspaceCostsFromBQ(workspacesPartition));
+      logger.info(String.format("Retrieved %d workspaces from BQ", liveCostsInBQ.size()));
```

Comment:
nit: Is this your debugging log or you think this might be useful? If I see this log in GAE, I would be confused about this.

Reply:
Agreed, I'll remove this line. This was for my debugging indeed